### PR TITLE
signature bridge contract implementation

### LINKED ISSUE #50: [TASK] Implement Signature Bridge Contract
Part of our protocol is the Signature Bridge implementation, which is developed as a contract in [protocol-cosmwasm](https://github.com/webb-tools/protocol-cosmwasm/tree/main/contracts/signature-bridge) and [protocol-solidity](https://github.com/webb-tools/protocol-solidity/blob/main/contracts/SignatureBridge.sol). The Signature Bridge is also implemented as a Pallet in [protocol-substrate](https://github.com/webb-tools/protocol-substrate/tree/main/pallets/signature-bridge).

### PR DESCRIPTION
**Pull request by dharjeezy**
closes #50 

### Task Description

- [x] Implemented Signature Bridge

- [x]  Make use of low level cross contract calls to handler contracts

- [x] Ink crates version bump across contracts and lib

- [x] Test for Admin Set Resource for Token wrapper handler in signature bridge

- [x] Test for Set fee proposal for Token wrapper handler in signature bridge

### CODE REVIEW

File: contracts/signature_bridge/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ark-crypto-primitives = { version = "^0.3.0", features = ["r1cs"], default-features = false }
+
+arkworks-setups = { version = "1.1.1",  features = ["r1cs"], default-features = false }
+arkworks-native-gadgets = { version = "1.0.0", default-features = false }
+
+[lib]
+name = "signature_bridge"
+path = "lib.rs"
+crate-type = [
+	# Used for normal contract Wasm blobs.
```
Comment (drewstone) [inline contracts/signature_bridge/Cargo.toml:30]: mixed tabs and spaces

File: contracts/signature_bridge/lib.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    pub fn element_encoder(v: &[u8]) -> [u8; 32] {
+        let mut output = [0u8; 32];
+        output.iter_mut().zip(v).for_each(|(b1, b2)| *b1 = *b2);
+        output
+    }
+
+    /// Get the `chain_id_type` from bytes array.
+    pub fn get_chain_id_type(chain_id_type: &[u8]) -> u64 {
```
Comment (drewstone) [inline contracts/signature_bridge/lib.rs:177]: Can we maybe add these functions to webb-rs and export them so we aren't re-using this so often?

**Review by drewstone — CHANGES_REQUESTED** [review]
Looks like a good start, can you make sure to add tests too?

File: contracts/signature_bridge/lib.rs (hunk 2 above)
Comment (dharjeezy) [reply contracts/signature_bridge/lib.rs:177]: Alright

File: contracts/signature_bridge/lib.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        buf[2..8].copy_from_slice(&chain_id_type);
+        u64::from_be_bytes(buf)
+    }
+
+    /// Computes the combination bytes of "chain_type" and "chain_id".
+    /// Combination rule: 8 bytes array(00 * 2 bytes + [chain_type] 2 bytes + [chain_id] 4 bytes)
+    /// Example:
+    ///    chain_type - 0x0401, chain_id - 0x00000001 (big endian)
+    ///    Result - [00, 00, 04, 01, 00, 00, 00, 01]
+    pub fn compute_chain_id_type(chain_id: u64, chain_type: &[u8]) -> u64 {
```
[the remark this answers is not in the document]
Comment (dharjeezy) [reply contracts/signature_bridge/lib.rs:189]: Alright

File: contracts/signature_bridge/Cargo.toml (hunk 1 above)
Comment (dharjeezy) [reply contracts/signature_bridge/Cargo.toml:30]: This was actually code generated, when i created the contract at first.

File: contracts/signature_bridge/lib.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct SignatureBridge {
+        governor: Vec<u8>,
+        proposal_nonce: u32,
+        chain_id: u64,
+        chain_type: [u8; 2],
+        counts: Mapping<Vec<u8>, [u8; 32]>,
+        resource_id_to_handler_address: Mapping<Vec<u8>, AccountId>,
```
[the remark this answers is not in the document]
Comment (dharjeezy) [reply contracts/signature_bridge/lib.rs]: > Looks like a good start, can you make sure to add tests too?

Yes, I will surely add tests. But for the signature bridge implementation to be complete, I need to have implemented the handlers specified here #54 #53 and #52

**Review by drewstone — CHANGES_REQUESTED** [review]

File: contracts/signature_bridge/lib.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        /// Invalid Resource Id
+        InvalidResourceId,
+    }
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct SignatureBridge {
+        governor: Vec<u8>,
+        proposal_nonce: u32,
+        chain_id: u64,
```
[the remark this answers is not in the document]
Comment (dharjeezy) [reply contracts/signature_bridge/lib.rs:45]: Alright. I get you now.

**Review by drewstone — APPROVED** [review]
LGTM!

**Comment by drewstone** [comment]
Yep, just good to fix it up imo.

**Pull request merged**